File: GFG/Stack/stockSpan.c
// import java.io.*;
// import java.lang.*;
// import java.util.*;

// #!/usr/bin/python -O

#include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
 #include<stdbool.h>
 #include<limits.h>

// # // #include<iostream>
// # // #include<algorithm>
// # // #include<string>
// # // #include<vector>
// # // using namespace std;

/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/Stacks
# Problem Statement	: Stock span problem
# Description		: 
# Complexity		: 
# =======================
# sample output
# ----------------------

# =======================
*/

struct stack 
{
	int data;
	struct stack *next;
	
};

struct stack *newNode(int data)
{
	struct stack *node =  (struct stack *) malloc(sizeof(struct stack));
	if(!node)
		return NULL;
	node->data =data;
	node->next = NULL;
	
	return node;
	
}

bool isEmpty(struct stack *s)
{
	return s==NULL;
}
void push(struct stack **s, int data)
{
	if(isEmpty(*s))
	{
		*s = newNode(data);
		
	}else{
		struct stack *t = newNode(data);
		t->next = *s;
		*s = t;
		
	}
	
}

int pop(struct stack **s)
{
	if(isEmpty(*s))
		return -1;
	struct stack *node = *s;
	*s = node->next;
	int d = node->data;
	free(node);
	return d;
	
	
}


int top(struct stack **s)
{
	
	if(isEmpty(*s))
		return -1;
	return (*s)->data;
	
}

void printArray(int arr[],  int n)
{
	printf("\n");
	int i;
	for(i=0; i<n ; i++)
		printf("%d ", arr[i]);
	
}

void calculateSP(int price[], int n ,int S[])
{
	S[0 ] = 1;
	
	struct stack *s = NULL;
	push(&s, 0);
	int i;
	for(i=0; i<n; i++)
	{
		while(!isEmpty(s) && price[top(&s)]<=price[i])
			pop(&s);
		
		S[i] = isEmpty(s)?i+1:i-top(&s);
		
		push(&s, i);
	}
	
}
int main()
{
	int price[] = {10, 4, 5, 90, 120, 80};
	int n  =  sizeof(price)/sizeof(price[0]);
	int S[n];
	calculateSP(price, n, S);
	
	printArray(S, n);
	
	
}

File: GFG/BT/StockSpan.java
import java.lang.*;
import java.util.*;
import java.io.*;

// #!/usr/bin/python -O

// #include <stdio.h>
// #include <stdlib.h>
// #include <string.h>
// #include<stdbool.h>
// #include<limits.h>

// #include<iostream>
// #include<algorithm>
// #include<string>
// #include<vector>
// #include<queue>
// using namespace std;

/*

# Author			: @RAJ009F
# Topic or Type 	: GFG/BT
# Problem Statement	:  Stock Span Problem
# Description		: 
# Complexity		: 
=======================
#steps:
=----------------


#sample output
----------------------

=======================
*/

class Node
	{
		int data;
		Node l;
		Node r;
		
		Node(int data)
			{
				this.data = data;
				this.l =  this.r = null;
			
			}
	
	}
	
class StockSpan	

{
	public void printStockSpan(int price[])
	{
		
		int n =  price.length;
		
		int[] S =  new int[n];
		
		Stack<Integer> s=  new Stack<Integer>();
		s.push(0);
		for(int i=1; i<n; i++)
		{
			while(!s.isEmpty() && price[s.peek()]<=price[i])
				s.pop();
			
			S[i] = s.isEmpty()?i+1:i-s.peek();
			s.push(i);
		}
		
		
		for(int i=0; i<n; i++)
			System.out.print(S[i] + " ");
		
	}
	
	public static void main(String args[])
	{
		
		StockSpan ss  =  new StockSpan();
		
		int price[] = {10,5,4,90,100,200,400,500,150,100,200};
			
		ss.printStockSpan(price);
		
		
	}








}